### PR TITLE
Add custom exceptions for CairoSerializer

### LINKED ISSUE #324: Wrap all the exceptions raised by StarkNet.py with custom ones.
Currently, StarkNet.py uses generic errors throughout the code.

This behaviour is not ideal because:
-  it doesn't allow for catching errors from this package separately,
- it doesn't differentiate between unexpected and expected errors.

The new exceptions could be introduced as a non-breaking change by using inheritance (possibly multiple inheritance).

### LINKED ISSUE #385: Document exceptions raised by methods
**Is your feature request related to a problem? Please describe.**

Where possible, we should include `::raises Exception` to method docstrings so it's explicit what exceptions are being raised where.

**Describe the solution you'd like**

as above

**Describe alternatives you've considered**

none

**Additional context**

n/a

### PR DESCRIPTION
**Pull request by war-in**
### **Please check if the PR fulfills these requirements**
- [ ] The formatter, linter, and tests all run without an error
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

## **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Closes #324 
Closes #385 

* **What is the current behavior?** (You can also link to an open issue here)

CairoSerializer raises ValueError and TypeError

## **What is the new behavior (if this is a feature change)?**

CairoSerializer raises CairoSerializerException and InvalidDataException

## **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

No

### CODE REVIEW

File: starknet_py/utils/data_transformer/data_transformer.py [hunk 1]
```diff
@@ -405,19 +418,25 @@ def to_python(self, value_types: List[dict], values: CairoData) -> NamedTuple:
...
         for name, cairo_type in type_by_name.items():
-            transformed, values = self.resolve_type(cairo_type).to_python(
-                cairo_type, name, values
-            )
+            try:
+                transformed, values = self.resolve_type(cairo_type).to_python(
+                    cairo_type, name, values
+                )
+            except ValueError as err:
+                raise InvalidDataException(str(err)) from err
```
Comment (cptartur) [inline starknet_py/utils/data_transformer/data_transformer.py:434]: Why was this added? If all errors were replaced by custom ones, are these lines necessary?

File: starknet_py/utils/data_transformer/data_transformer.py [hunk 2]
```diff
@@ -405,19 +418,25 @@ def to_python(self, value_types: List[dict], values: CairoData) -> NamedTuple:
...
+
             result[name] = transformed
 
         if len(values) > 0:
-            raise ValueError(
+            raise CairoSerializerException(
```
Comment (cptartur) [inline starknet_py/utils/data_transformer/data_transformer.py:439]: This could probably also be `InvalidDataException`.

File: starknet_py/utils/data_transformer/data_transformer.py [hunk 3]
```diff
@@ -169,12 +175,12 @@ def from_python(self, cairo_type, name, value) -> CairoData:
             return [value & ((1 << 128) - 1), value >> 128]
 
         if not isinstance(value, dict):
-            raise TypeError(f"Expected {name} to be a dict.")
+            raise InvalidDataException(f"Expected {name} to be a dict.")
```
Comment (cptartur) [inline starknet_py/utils/data_transformer/data_transformer.py:178]: This is a breaking change

File: starknet_py/utils/data_transformer/data_transformer.py (hunk 1 above)
Comment (war-in) [reply starknet_py/utils/data_transformer/data_transformer.py:434]: Yes, because not all of them were replaced. Felt still returns ValueError, because I didn't want to change any errors outside the module

File: starknet_py/utils/data_transformer/data_transformer.py (hunk 3 above)
Comment (war-in) [reply starknet_py/utils/data_transformer/data_transformer.py:178]: You're right, I'll return to the old TypeError here because it is catched in the CairoSerializer.from_python

File: starknet_py/utils/data_transformer/data_transformer.py (hunk 1 above)
Comment (cptartur) [reply starknet_py/utils/data_transformer/data_transformer.py:433]: Are we sure only "default" exception that can be thrown in this module is `ValueError`? Perhaps this should be general `except Exception as err`?

File: starknet_py/utils/data_transformer/data_transformer.py (hunk 3 above)
Comment (cptartur) [reply starknet_py/utils/data_transformer/data_transformer.py]: Can't we just throw an exception inheriting a proper type here? That's way it's not breaking.

Comment (war-in) [reply starknet_py/utils/data_transformer/data_transformer.py]: It is CairoSerializerException and I would like avoid to use it outside CairoSerializer

File: starknet_py/utils/data_transformer/data_transformer.py (hunk 1 above)
Comment (war-in) [reply starknet_py/utils/data_transformer/data_transformer.py:433]: Yes, it is only the ValueError. Other errors are general python errors and we shouldn't catch them

File: starknet_py/utils/data_transformer/data_transformer.py [hunk 4]
```diff
@@ -138,7 +142,7 @@ def from_python(self, cairo_type, name, value) -> CairoData:
             return [value]
 
         if not isinstance(value, int):
-            raise TypeError(f"{name} should be int.")
```
Comment (cptartur) [inline starknet_py/utils/data_transformer/data_transformer.py:141]: This is also a breaking change

File: starknet_py/utils/data_transformer/data_transformer.py [hunk 5]
```diff
@@ -386,11 +396,14 @@ def from_python(
...
 
-            values = self.resolve_type(cairo_type).from_python(
-                cairo_type, name, named_arguments[name]
-            )
+            try:
+                values = self.resolve_type(cairo_type).from_python(
+                    cairo_type, name, named_arguments[name]
+                )
+            except (TypeError, ValueError) as err:
+                raise InvalidDataException(str(err)) from err
```
Comment (cptartur) [inline starknet_py/utils/data_transformer/data_transformer.py:406]: This is still kind of problematic, because some `TypeErrors` would be that way replaced by `InvalidDataException` which extends `ValueError`

File: starknet_py/utils/data_transformer/data_transformer_test.py [hunk 6]
```diff
@@ -592,7 +596,7 @@ def test_wrong_types(cairo_type, value):
         }
     ]
 
-    with pytest.raises(TypeError):
```
Comment (cptartur) [inline starknet_py/utils/data_transformer/data_transformer_test.py:595]: This should raise `CairoSerializerException`.

**Review by cptartur — CHANGES_REQUESTED** [review]

**Review by cptartur — APPROVED** [review]

File: starknet_py/utils/data_transformer/errors.py [hunk 7]
```diff
@@ -0,0 +1,6 @@
+class CairoSerializerException(TypeError):
+    """Exception thrown by CairoSerializer"""
+
+
+class InvalidDataException(ValueError):
```
Comment (Solpatium) [inline starknet_py/utils/data_transformer/errors.py:5]: So this is an error for `to_python` methods? Maybe we could name it `DeserializationError` and the other one `SerializationError`?

File: starknet_py/utils/data_transformer/data_transformer.py [hunk 8]
```diff
@@ -386,11 +396,16 @@ def from_python(
...
-                cairo_type, name, named_arguments[name]
-            )
+            try:
+                values = self.resolve_type(cairo_type).from_python(
+                    cairo_type, name, named_arguments[name]
+                )
+            except ValueError as err:
+                raise InvalidDataException(str(err)) from err
+            except TypeError as err:
+                raise CairoSerializerException(str(err)) from err
```
Comment (Solpatium) [inline starknet_py/utils/data_transformer/data_transformer.py:408]: What kind of exceptions do we expect here?

File: starknet_py/utils/data_transformer/errors.py (hunk 7 above)
Comment (Solpatium) [inline starknet_py/utils/data_transformer/errors.py:5]: I would add an abstract class to inherit from that would mean any error coming from data_transformer. This way protostar can expect this one error.

Comment (Solpatium) [reply starknet_py/utils/data_transformer/errors.py:5]: I see that it is not the case. Users won't understand their meaning.

File: starknet_py/utils/data_transformer/data_transformer.py (hunk 8 above)
Comment (war-in) [reply starknet_py/utils/data_transformer/data_transformer.py:408]: I'm not sure if I understand you correctly 😅 We expect errors from transformer classes (TupleTransformer, FeltTransformer). I've changed errors in the data_transformer.py, but I don't want to do it in e.g. Felt.py (it throws ValueErrors)

File: starknet_py/utils/data_transformer/data_transformer_test.py [hunk 9]
```diff
@@ -617,7 +620,7 @@ def test_arg_provided_twice():
 def test_missing_arg():
     abi = [{"name": "first", "type": "felt"}, {"name": "second", "type": "felt"}]
 
-    with pytest.raises(TypeError) as excinfo:
+    with pytest.raises(CairoSerializerException) as excinfo:
```
Comment (Solpatium) [inline starknet_py/utils/data_transformer/data_transformer_test.py:623]: We should still catch exact exceptions here, not the general one

File: starknet_py/utils/data_transformer/data_transformer.py (hunk 8 above)
Comment (Solpatium) [reply starknet_py/utils/data_transformer/data_transformer.py]: Those exceptions might not be very informative for users, but we can keep it this way.

**Review by Solpatium — APPROVED** [review]
Just one comment
